### DISCUSSION TITLE
Truncating blanks in media files in order to achieve better crossfades

### DISCUSSION THREAD

**Discussion by SpeedyCharly** [AzuraCast/AzuraCast#6343] [General]
Following @Moonbase59 suggestion, let's discuss ways of truncating/skipping leading/ending blanks in media files so that crossfades are truly that and don't happen over dead air.  
In Sam Broadcaster (and probably other streaming apps), leading/ending blanks are truncated prior to crossfade resulting in smooth and predictable crossfades. Such a function doesn't exist in LS (hint-hint @Toots LOL) but after a bit of 'off-topic' discussion with Moobase59 [ https://github.com/AzuraCast/AzuraCast/issues/6327 ] I've learner that a bit of custom coding in LS config can possibly achieve that (albeit at some CPU usage cost) using _blank.eat_ and _blank.skip_ operators. 
Not being a Linuxian geek myself (just a knuckle-headed radio operator actually), I can barely grasp the concept. So I'm starting this discussion, hoping that we can butt heads and come up with the slickest possible bit of custom coding to get the job done.
In Moonbase's words; "You can use a combination of [blank.eat(at_beginning=true,…)](https://www.liquidsoap.info/doc-dev/reference.html#blank.eat) and [blank.skip(…)](https://www.liquidsoap.info/doc-dev/reference.html#blank.skip) to achieve just that. Just remember blank.eat(…) is costly."

So, anyone interested in developing this idea???

@Toots also suggested preformatting media files prior to uploading to server so that such coding wouldn't be necessary, but I have yet to find an app that can do a decent batch truncate job. With my next to 300,000 files media collection, manual truncates would take me into the next century...

**Comment by Moonbase59** [comment]
Just some random thoughts, because I’m in a hurry…

- In https://github.com/AzuraCast/AzuraCast/discussions/6334, I forgot to mention you’ll need a rather new rolling release AzuraCast version, and one that has Liquidsoap 2.2.x, because the way AzuraCast annotates songs has changed (to the better).

- I currently have "only" 156k songs, of which rather few are recorded from my old LPs and might have some silence _at the beginning_. I _do_ keep these odd songs with _<song> <silence> <bonus track>_ I ripped from my CDs, because I consider my central FLAC music collection a bit "sacred" on don’t allow anything that isn’t in congruence with _MusicBrainz_ (I use MusicBrainz Picard plus a custom plugin for tagging).

- I have also decided to host AzuraCast at home and stream to a _relay_ in the cloud. This allows for better maintenance and keeping my whole collection automatically in sync. AzuraCast has an NFS mount to that, and can only access it in _read-only_ mode, for added safety.

- I currently only use the so-called _simple_ crossfading, because the "smart" variant sometimes produces wild loudness jumps.

- My collection is completely ReplayGain-tagged using [_loudgain_](https://github.com/Moonbase59/loudgain), so I can skip the extra overhead of letting Liquidsoap/AzuraCast calculate ReplayGain (which is costly in terms of CPU). Henceforth, I switched ReplayGain processing in AzuraCast _off_ and simply use the existing tags.

- My station will cut songs with silence _at the end_ and such containing "bonus tracks" after the first song, in order not to broadcast silence.

- I currently _don’t_ use `blank.eat()` to remove silence from the beginning of tracks, but will try my own suggestion in the next days. SAM Broadcaster actually _shines_ in some respects, but it’s Windows-only and we want FOSS, after all.

- My only addition to the Liquidsoap config (apart from setting up a video stream, which is something entirely different) is currently this:

```
# Be sure to have ReplayGain applied before crossing.
radio = amplify(1.,override="replaygain_track_gain",radio)

# If a song has <song> <silence> <bonus track>,
# blank.eat() will skip <silence> and play <bonus track>
#radio = blank.eat(radio, max_blank=5.0, threshold=-60.0)
# blank.skip() will skip to next song and NOT play <bonus track>.
# Defaults: max_blank=20.0 (seconds), threshold=-40.0 (dB)
radio = blank.skip(radio, max_blank=5.0, threshold=-70.0)

def live_aware_crossfade(old, new) =
    if to_live() then
        # If going to the live show, play a simple sequence
        sequence([fade.out(old.source),fade.in(new.source)])
    else
        # Otherwise, use the smart transition
        log("Using custom crossfade")
        cross.simple(old.source, new.source, fade_in=0.00, fade_out=3.50)
    end
end

radio = cross(minimum=0., duration=4.00, width=2.00, live_aware_crossfade, radio)
```

- For jingles, bumpers and sweepers I’ve set up the cut points, fading and overlap by editing each in AzuraCast, often with just fade-in, fade-out and overlap set to `0.1` since `0` seems to ignore that parameter.

- I’m still thinking of modifying the custom crossfade above to handle jingles specially, this should (hopefully) be possible by checking the `jingle_mode` annotation.

So much for now. Have a good time & goodnight!

**Reply by SpeedyCharly** [reply]
@Moonbase59 
Great... you've touched on most of the things I've discussed in the past regarding getting cleaner crossfades in LS.
- Jingles and short media (90 sec or less) should NOT be crossfaded. Furthermore, jingles and such should not increment the media count to afford us better/more predictable control of the playlist clockwheel. The _jingle_mode_ tag should also be useful there (but that's another topic altogether)
- Truncating leading and ending silence is a big concern and should be 1st priority. But good preprocessing might prove to be a better solution rather than adding more blades to the LS 'swiss army knife', saving precious CPU resources for other LS processes. We need to find a good batch silence truncator (hopefully FOSS).  Am currently testing **[Audio Silence Trimmer Pro](http://www.dandans.com/Audio-Silence-Trimmer-Pro.htm)** (proprietary $$$), but noted some caveats in the first few test runs...
- _ReplayGain_ which has a bearing on the crossfade process should IMHO occur before crossfade.  Again, normalizing before uploading media to LS may be the best way to go to save CPU resources.  At first glance, your **loudgain** seems too 'geeky' for most of us lowly radio operators, so I'll stick with my **[Qmp3Gain](https://sourceforge.net/projects/qmp3gain/files/qmp3gain/0.9.3/)** for the time being (same engine under the hood anyhow)...
- I like your custom crossfade routine with 3 secs. fade-out and 0 fade-in (default in Sam Broadcaster) which should be hardwired into LS/AzuraCast, together with comprehensive knobs and sliders to make it easily customizable for different music environments/operator preferences. >> Do you have to disable crossfade in LS in order to use this custom cross??

Don't want to make this a tome of biblical proportions, so I'll end this here by saying I totally agree with you that it's a shame that a brilliant streaming engine like **Sam Broadcaster** it's available on **Apple** or **Linux**...  Years ago, before the advent of AutoDJ, I used SamB on a local machine to feed an OutCast radio server...  Good ole days ...

**Reply by Moonbase59** [reply]
> Do you have to disable crossfade in LS in order to use this custom cross??

I distinctly remember there was some issue when setting _Edit Station Profile → AutoDJ → Crossfade Method_ to "Disable Crossfade" in an older rolling release, so I set it to "Normal Mode" (i.e., simple crossfade) and 2s and trust my code will overwrite that… (which only works in newer rolling releases).

It _does_ introduce duplicate crossfading code within the Liquidsoap Config, but somehow our code seems to "win" if put into the input box just above the `def live_aware_crossfade(old, new) =` code.

Maybe, whenever I’ve got some time, I should retry with station crossfading set to "Disable Crossfading"… If only the day had 48 hours… ;-)

**Comment by Moonbase59** [comment]
Okay, since mass-stripping silence (at start and end, _not_ within the song) might be of interest to some, I thought using the Swiss-Army-knife of audio and video processing, `ffmpeg` will be the best bet and wrote a little bash script that will do exactly that, for FLAC, MP3, WAV, OGG, OGA and WMA audio files:

```bash
#!/usr/bin/bash

# strip-silence
# 2023-06-15 - Moonbase
# strips silence at beginning and end of an audio file
# Put this in your `~/bin` or `/usr/local/bin` folder for easy access.
# You might need to set the execute bit (`chmod +x strip-silence`).

# BEWARE! This is not exact science, and you might easily mess up your valuable
# audio files! Be sure to make BACKUPS of you music first, then try with a small
# and representative set of files. Adjust THRESHOLD, KEEP_SILENCE and SUFFIX
# to your needs.
#
# Note that we need to re-encode, so the generated files (except for
# FLAC and WAV) will suffer from GENERATION LOSS!
#
# Note also that ffmpeg isn’t too good in keeping METADATA, so these should be
# checked after stripping.
#
# Adjust extra encoding parameters and add additional file types
# in the case statement below, following the examples given.

THRESHOLD="-70dB" # adjust to your needs
NONSILENCE=0 # how many seconds of non-silence before it stops trimming
KEEP_SILENCE=0 # seconds of silence to keep (at both ends)
SUFFIX="_stripped" # suffix to add to the filename
OVERWRITE="-y" # specify "-y" to force overwriting existing output files, "" otherwise

# The actual ffmpeg filter command
# Note that we reverse the audio to remove silence at the end,
# in order to avoid removing anything in the middle of the song.
FFMPEG_CMD="-map_metadata 0 -af silenceremove=start_periods=1:start_duration=${NONSILENCE}:start_silence=${KEEP_SILENCE}:start_threshold=${THRESHOLD}:detection=peak,aformat=dblp,areverse,silenceremove=start_periods=1:start_duration=${NONSILENCE}:start_silence=${KEEP_SILENCE}:start_threshold=${THRESHOLD}:detection=peak,aformat=dblp,areverse"


while [ "$1" != '' ] ; do

  fname=${1%.*}
  fext=${1##*.}
  fext=${fext,,}

  # since re-encoding suffers from generation loss, set desired parameters per file type
  case "$fext" in
  
    "flac")
      ffmpeg -hide_banner -i "$1" $FFMPEG_CMD $OVERWRITE "$fname$SUFFIX.$fext"
      ;;
    
    "mp3")
      ffmpeg -hide_banner -i "$1" $FFMPEG_CMD -ab 320k $OVERWRITE "$fname$SUFFIX.$fext"
      ;;
    
    "wav")
      ffmpeg -hide_banner -i "$1" $FFMPEG_CMD $OVERWRITE "$fname$SUFFIX.$fext"
      ;;
    
    "ogg" | "oga")
      ffmpeg -hide_banner -i "$1" $FFMPEG_CMD $OVERWRITE "$fname$SUFFIX.$fext"
      ;;
    
    "wma")
      ffmpeg -hide_banner -i "$1" $FFMPEG_CMD $OVERWRITE "$fname$SUFFIX.$fext"
      ;;
    
    # none of the above
    *)
      echo "$1 is not a recognized file type, skipping…"
      ;;
  esac
  
  shift
done
```

You will find the script and some (royalty-free) test files in the attached ZIP:
[testfiles-strip.zip](https://github.com/AzuraCast/AzuraCast/files/11757008/testfiles-strip.zip)

The `file-with-3s-silence.*` files have 3s of silence, 3s of test tone, and another 3s of silence each. The song file has what _looks like_  2.5s silence at the start and end, but _isn’t really_. At -70 dB threshold, `strip-silence` should remove about 1s at both ends.

The `bash` will do all globbing for you, so you can easily use something like:
```bash
strip-silence ~/Music/*.flac
strip-silence *.flac *.wav *.mp3
```
or even
```
strip-silence *
```
which will only strip files with extensions listed in the `case` statement.

This is for Linux (may work on MacOS, too). Windows users can use WSL bash, or write their own Powershell script.

Example file `file-with-3s-silence.flac` before stripping:
![Auswahl_221](https://github.com/AzuraCast/AzuraCast/assets/3706922/481aa593-6e18-4c20-99dd-1feaafd1b920)

After `strip-silence file-with-3s-silence.flac` with `KEEP_SILENCE=0`:
![Auswahl_222](https://github.com/AzuraCast/AzuraCast/assets/3706922/7cc87839-099f-4497-bdec-b84c0de21e67)

After `strip-silence file-with-3s-silence.flac` with `KEEP_SILENCE=1`:
![Auswahl_223](https://github.com/AzuraCast/AzuraCast/assets/3706922/6c0d6e7e-ceb9-4ce4-af9e-0a4e559e31e1)


**Please BACKUP your files before trying this!**
**Carefully read and follow the comments at the beginning of the script, and modify to your needs.**

No guarantees. This is just to show how we can do things using Free and Open Source Software, instead of buying costly software that probably won’t even be as flexible as this.

I’ll leave it as an exercise for the interested reader to modify the script to recursively scan whole folders (hint: use `find`) if given as an argument. :-)

**Reply by BusterNeece** [reply]
@Moonbase59 Wow, nicely done. We're looking into possibly having an "advanced import" dialog box that allows you to pre-calculate Replaygain values, trim silence, and possibly convert media formats automatically during the import process. We may end up using something very similar to this.

**Reply by Moonbase59** [reply]
Thanks :-) I really love ffmpeg, it can do almost _anything_… You can use the filter in any case, if you like.

> We're looking into possibly having an "advanced import"

Oops. Nice idea. But that means I’ll probably have to bring [_loudgain_](https://github.com/Moonbase59/loudgain) up to standards, make it compilable again, and possibly set up a CI (which I don’t really know how to do). After literally years of neglecting it… If only the day had 48 hours… ;-) Wouldn’t want to replaygain with anything else, because loudgain handles so many filetypes (even video) and does everything right _mp3gain_ did wrong. (That’s why I wrote it, back in time.)

Please don’t forget an option not to touch anything, for audio maniacs like me, that even mount the music read-only, just to avoid corruption. Thanks!

**Reply by BusterNeece** [reply]
Of course, the default would be to just import items untouched. On our end, that's the most performant option, after all. Everything else would be optional add-ons.

So how does Liquidsoap do replaygain calculation then? We may end up writing little Liquidsoap scripts and executing those so we can reuse our existing generative code (and abstract out our library dependencies) to do the processing.

**Reply by Moonbase59** [reply]
Good question, we should ask @toots or @smimram maybe. I _think_ they used ffmpeg in olden days but now switched to something internal. (I hope they used some loudgain code, since loudgain is even mentioned in the Liquidsoap book, as I recently discovered. Thanks for that!)

**Comment by Moonbase59** [comment]
Addendum: **QUICK ’N’ DIRTY: The globstar**

If you don’t want to fiddle around with programming and _still_ wish to recursively strip files in certain (sub-)folders, you can activate the _globstar_ in your `bash` and do just that.

Be aware that globbing fills the argument list, which is finite in size (use `getconf ARG_MAX` to check)! That is the reason why the ability to use `**` is usually _disabled_ in newer bash versions. Using `**` will allow to recursively traverse through all subfolders, looking for the files you specify.

So if you
```bash
$ shopt -s globstar
```
and then try to list all FLAC files in your collection
```bash
$ ls ~/Musik/Tagged/**/*.flac 
```
you’ll most certainly get
```
bash: /usr/bin/ls: The argument list is too long
```
but if you reduce yourself to all AC/DC FLAC files, you might just succeed:
```bash
$ ls ~/Musik/Tagged/AC_DC/**/*.flac
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Back in Black.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Givin'\'' the Dog a Bone.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Have a Drink on Me.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Hells Bells.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Let Me Put My Love Into You.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Rock and Roll Ain'\''t Noise Pollution.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Shake a Leg.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - Shoot to Thrill.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - What Do You Do for Money Honey.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Back in Black (2005 album, US)/AC_DC - You Shook Me All Night Long.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Are You Ready.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Back in Black.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Bonny.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Dirty Deeds Done Dirt Cheap.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Fire Your Guns.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - For Those About to Rock (We Salute You).flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Heatseeker.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Hells Bells.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - High Voltage.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Highway to Hell.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Jailbreak.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Let There Be Rock.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Moneytalks.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Shoot to Thrill.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Sin City.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - T.N.T..flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - That'\''s the Way I Wanna Rock '\''n'\'' Roll.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - The Jack.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - The Razors Edge.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Thunderstruck.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Who Made Who.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - Whole Lotta Rosie.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - Live (2003 album, live, XE)/AC_DC - You Shook Me All Night Long.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Code Red.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Demon Fire.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Kick You When You'\''re Down.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Money Shot.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - No Man'\''s Land.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Realize.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Rejection.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Shot in the Dark.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Systems Down.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Through the Mists of Time.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Wild Reputation.flac'
'/home/matthias/Musik/Tagged/AC_DC/AC_DC - POWER UP (aka PWR'$'\342\232\241\357\270\216''UP) (2020 album, XE)/AC_DC - Witch'\''s Spell.flac'
```

Instead of `ls`, you could use `strip-silence` as well, but I never told you…

**Comment by SpeedyCharly** [comment]
@ Moonbase59

> Oops. Nice idea. But that means I’ll probably have to bring [loudgain](https://github.com/Moonbase59/loudgain) up to standards, make it compilable again, and possibly set up a CI (which I don’t really know how to do). After literally years of neglecting it… If only the day had 48 hours… ;-) Wouldn’t want to replaygain with anything else, because loudgain handles so many filetypes (even video) and does everything right mp3gain did wrong. (That’s why I wrote it, back in time.)

Us knuckle-heads would love a Windows .exe and an Apple installer of **loudgain**  ! Pretty please LOL

@BusterNeece 

> We're looking into possibly having an "advanced import" dialog box that allows you to pre-calculate Replaygain values, trim silence, and possibly convert media formats automatically during the import process

How's about creating a stand-alone app that does all of that + normalize as a side project?
